src/main/java/net/objecthunter/exp4j/tokenizer/Tokenizer.java

```diff
@@ -154,6 +179,12 @@ private Token parseFunctionOrVariable() {
                 if (f != null) {
                     lastValidLen = len;
                     lastValidToken = new FunctionToken(f);
+                } else if (undefinedVariables != null) {
+                	if (! undefinedVariables.contains(name)) {
```

Comment:
If undefinedVariables is declared Collection the check should be there, since we cannot assume the check needn't be done. If it is declared Set, we can omit the check. I usually don't care about these kind of optimisations, for sets that are small.